### ISSUE TITLE
Elastic Endpoint v8.7 Fanotify permission checks possible deadlock tainting kernel in Debian

### ISSUE THREAD

**Issue by gpapakyriakopoulos** [elastic/endpoint#47]
As the title suggests, it seems that when using the Elastic Defend integration with elastic-agent on version 8.7 on Linux Debian Buster, it taints the kernel with the following message from ElasticEndpoint : 

```
kernel: [34746048.630455] elastic-endpoin (30072): Using fanotify permission checks may lead to deadlock; tainting kernel
```

This also seems to be leading to potential deadlock states in the Linux kernel, which subsequently generates messages like the following : 

```
kernel: [  363.761784] INFO: task elastic-endpoin:537 blocked for more than 120 seconds.
```

This behaviour seems to have been introduced in the Debian built by the following patch : https://scm.linefinity.com/debian/linux/commit/014289b45954c7a334f25d9dbf224b120430fb5a

`Elastic Agent version : 8.7.0`
`Linux kernel version :  4.19.0-18-amd64 #1 SMP Debian 4.19.208-1 (2021-09-29) x86_64 GNU/Linux`

Disabling the `Malware Protection` from the Elastic Defend integration fixes the issue.

**Comment by nicholasberlin** [comment]
Thanks for reporting. We will get back to you.

**Comment by nicholasberlin** [comment]
What file systems are mounted during the deadlock messages?

**Comment by gpapakyriakopoulos** [comment]
Here is an example from one of our hosts that has been affected by the issue (not all affected hosts had the same FSs mounted) :

```
mount | grep / | awk '{ print $5 }'
sysfs
proc
devtmpfs
devpts
tmpfs
ext4
securityfs
tmpfs
tmpfs
tmpfs
cgroup2
cgroup
pstore
bpf
cgroup
cgroup
cgroup
cgroup
cgroup
cgroup
cgroup
cgroup
cgroup
cgroup
hugetlbfs
autofs
debugfs
mqueue
ext4
binfmt_misc
tracefs
```

**Comment by nicholasberlin** [comment]
Are they all running the same kernel version?

**Comment by gpapakyriakopoulos** [comment]
Did a quick check, we have a mix of kernel versions, the same issue also appeared on : 
* `4.19.0-14-amd64 #1 SMP Debian 4.19.171-2 (2021-01-30) x86_64`
* `5.10.0-15-amd64 #1 SMP Debian 5.10.120-1 (2022-06-09) x86_64 GNU/Linux`

**Comment by nicholasberlin** [comment]
Oh, for the 5.10 kernel, were there deadlock messages or just the kernel taint message?

**Comment by gpapakyriakopoulos** [comment]
As far as I can tell it is only the kernel taint message on that one. In any case, the actual deadlocks only happened in 2 hosts that were under much heavier load than the rest of the affected ones and unfortunately neither of those is 5.X kernel.

There might be a correlation here, but not sure if there is causation.

**Comment by nicholasberlin** [comment]
We use a different permission flag starting in 5.0 kernels;`FAN_OPEN_EXEC_PERM` vs `FAN_OPEN_PERM`.
Working to reproduce. Thanks again.

**Comment by nicholasberlin** [comment]
I've been building Linux kernel in a loop without issue :(

I'm running
```
Linux version 4.19.0-14-amd64 (debian-kernel@lists.debian.org) (gcc version 8.3.0 (Debian 8.3.0-6)) #1 SMP Debian 4.19.171-2 (2021-01-30)
```
in a VM with two CPUs.

Can you describe what the system's load is?
Is there a lot of containers/mount activity?

Thanks.

**Comment by gpapakyriakopoulos** [comment]
Interesting, some systems are heavy on containers/mount activity as you say. Other systems that also got a tainted kernel are less load-heavy, I can share details if you point me to what details would make more sense.

**Comment by gpapakyriakopoulos** [comment]
Hey @nicholasberlin sorry took too long to respond. It does seem that systems these happens into are mostly systems with a lot of containers coming up and down as you say, so it is most likely related to that. 

Since we also have the kernel taint part, we've opted to disable the malware protection part of the elastic-agent policy for now.

At of curiosity about the Fanotify part, is there a particular reason elastic-agent and Elastic Endpoint have opted to not use kernel modules at all as a primary means of tracking files/syscalls and are instead relying solely on Fanotify. Obviously using native Kernel APIs is better than compiling and injecting custom kernel modules in a theoretical sense, but since it seems the API you are using was not intended to be used that way (for whatever reason they decided that), wouldn't it make sense to have a kernel module as a means of achieving the same ?

**Comment by nicholasberlin** [comment]
> Since we also have the kernel taint part, we've opted to disable the malware protection part of the elastic-agent policy for now.

Understandable. I hope you will re-attempt with 8.7.1 (or newer) in the future.

> Out of curiosity about the Fanotify part, is there a particular reason elastic-agent and Elastic Endpoint have opted to not use kernel modules at all as a primary means of tracking files/syscalls and are instead relying solely on Fanotify ?

Generally, to allow for the largest possible install base. Practically, because it's easier and safer.

1. Easier: userspace-only side steps a lot of logistical issues
- Does Elastic maintain compiled modules for all kernels we support? This will exclude users with custom kernels and unsupported distros. What about supported distros with new kernels? Users will have to consider Endpoint's kernel support when scheduling upgrades.
- Does Endpoint instead compile the modules on the host? What if the infrastructure to compile is not present or explicitly not allowed on the host? What if the compilation fails for a new kernel?
- Attempt to push a kernel module/driver to mainline? Bug fixes would be on Linux's timeline. Old kernels/distros would not benefit.

2. Safer: kernel module bugs are worse than userspace bugs
- Admittedly, `fanotify` can cause systemic issues if used improperly, but the surface area for bugs is low compared to a kernel module where any coding mistake could cause a box to go down. A bug in a kernel module is likely far more severe than a bug in `fanotify` handling userspace code.

> Obviously using native Kernel APIs is better than compiling and injecting custom kernel modules in a theoretical sense, but since it seems the API you are using was not intended to be used that way (for whatever reason they decided that), wouldn't it make sense to have a kernel module as a means of achieving the same ?

According to the [docs](https://man7.org/linux/man-pages/man7/fanotify.7.html) for `fanotify`: `Use cases include virus scanning`.  In fairness, Debian's issue is specifically with the `PERM` flags, not `fanotify` in general. However, if Endpoint used non-`PERM` flags to avoid the kernel taint, Endpoint would no longer be *preventing* malware from executing. It would be racing to stop something which has already begun executing.

Assuming the kernel module issues mentioned above were resolved, writing a kernel module to achieve these ends would effectively be rewriting existing kernel functionality. Functionality which has been around since Linux v2.6.31. Even with a more targeted implementation, it seems foolish to think we could do a better (or even equivalent) job than the years of production testing the system has gone through. (Just my opinion.)

In general, `fanotify` gives a lot of power to user space and can be problematic. So, I can understand why Debian added the taint flag and message. This particular taint flag is less severe than other taint flags, however. For example, an unsigned kernel module taint flag is basically saying all bets are off because new code has been injected into the kernel and there are no limits to what that new code can do, including completely changing the kernel itself.  Tainting the kernel because userspace is using `fanotify` permission flags is not saying anything about the integrity of the kernel. It is saying the kernel is now dependent on userspace code, but the kernel itself is unchanged.

If you trust Elastic Endpoint, then I believe, this particular taint flag and message can be ignored. Obviously, messages like:
```
kernel: [  363.761784] INFO: task elastic-endpoin:537 blocked for more than 120 seconds.
```
are not ignore-able, and in this case, the taint message helped. But the taint message alone is not a reason to avoid Endpoint. That said, I can also understand wanting to avoid any and all kernel tainting situations.

**Comment by nicholasberlin** [comment]
Update: we identified an issue with our signal handling code, especially as it relates to segmentations faults, which can cause the issue identified above. A fix has been merged to our eventual 8.9 release with back ports to 7.17 and 8.8 branches working their way through CI.

**Comment by gpapakyriakopoulos** [comment]
Much appreciated and thanks for the very quick turnaround. We will make sure to test it when we are able to update to 8.8. I will close this to free-up your board since there is no further action to be had here.

Thanks again and have a good day!
